contracts/core/src/token/OmniExchange.sol

```diff
@@ -0,0 +1,209 @@
+// SPDX-License-Identifier: GPL-3.0-only
+pragma solidity 0.8.24;
+
+import { OwnableUpgradeable } from "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import { XAppUpgradeable } from "src/pkg/XAppUpgradeable.sol";
+import { FeeOracleV1 } from "src/xchain/FeeOracleV1.sol";
+import { ConfLevel } from "src/libraries/ConfLevel.sol";
+import { OmniFund } from "./OmniFund.sol";
+
+contract OmniExchange is XAppUpgradeable, OwnableUpgradeable {
+    event MaxSwapSet(uint256 max);
+    event FundSet(address fund);
+    event PctCutSet(uint256 pct);
+    event Swapped(address indexed recipient, uint256 paid, uint256 fee, uint256 cut, uint256 amtOMNI);
+
+    /**
+     * @notice Gas limit passed to OmniFund.tryWithdrawRemaining xcall
+     */
+    uint64 internal constant WITHDRAW_XCALL_GAS = 100_000;
+
+    /**
+     * @notice Denominator for pct cut
+     */
+    uint256 internal constant PCT_CUT_DENOM = 1000;
+
+    /**
+     * @notice Address of OmniFund on Omni
+     */
+    address public fund;
+
+    /**
+     * @notice Max amt (in native token) that can be swapped in a single tx
+     * @dev TODO: Should we denominate this in OMNI?
+     */
+    uint256 public maxSwap;
+
+    /**
+     * @notice Percentage cut (over PCT_CUT_DENOM) taken by this contract for each swap
+     *         Used to disencentivize spamming
+     */
+    uint256 public pctCut;
+
+    /**
+     * @notice Map address to total historical swap requests.
+     * @dev On each swap(), the total historical amount swapped for an account is sent to OmniFund.
+     *      This allows OmniFund to recover from failed transfers with retries. And mitigates reorg risk
+     *      on ConfLevel.Latest, as accounts that receive OMNI for ETH deposits that do not remain in
+     *      the canonical source will not receive OMNI for additional deposits, until total deposis
+     *      are accounted for.
+     */
+    mapping(address => uint256) public swapped;
+
+    constructor() {
+        _disableInitializers();
+    }
+
+    function initialize(address fund_, address portal, address owner) external initializer {
+        _setFund(fund_);
+        _setMaxSwap(0.02 ether);
```

Comment:
should avoid hardcoded params like this, since changing the default will change the bytecode